File: pkg/zeromcp/schema.go
package zeromcp

import "fmt"

// SimpleType is a shorthand type name: "string", "number", "boolean", "object", "array".
type SimpleType = string

// InputField can be a SimpleType string or an ExtendedField for more detail.
type InputField struct {
	Type        SimpleType `json:"type"`
	Description string     `json:"description,omitempty"`
	Optional    bool       `json:"optional,omitempty"`
}

// Input maps field names to their type definitions.
// Use String/Number/Boolean/etc helpers, or InputField structs directly.
type Input map[string]any

// JsonSchema is a JSON Schema object for tool input validation.
type JsonSchema struct {
	Type       string                        `json:"type"`
	Properties map[string]JsonSchemaProperty  `json:"properties"`
	Required   []string                      `json:"required"`
}

// JsonSchemaProperty describes a single property in a JSON Schema.
type JsonSchemaProperty struct {
	Type        string `json:"type"`
	Description string `json:"description,omitempty"`
}

var typeMap = map[string]string{
	"string":  "string",
	"number":  "number",
	"boolean": "boolean",
	"object":  "object",
	"array":   "array",
}

// ToJsonSchema converts a simplified Input definition to a full JSON Schema.
func ToJsonSchema(input Input) JsonSchema {
	schema := JsonSchema{
		Type:       "object",
		Properties: make(map[string]JsonSchemaProperty),
		Required:   []string{},
	}

	if input == nil {
		return schema
	}

	for key, value := range input {
		switch v := value.(type) {
		case string:
			mapped, ok := typeMap[v]
			if !ok {
				panic(fmt.Sprintf("unknown type %q for field %q", v, key))
			}
			schema.Properties[key] = JsonSchemaProperty{Type: mapped}
			schema.Required = append(schema.Required, key)

		case InputField:
			mapped, ok := typeMap[v.Type]
			if !ok {
				panic(fmt.Sprintf("unknown type %q for field %q", v.Type, key))
			}
			prop := JsonSchemaProperty{Type: mapped}
			if v.Description != "" {
				prop.Description = v.Description
			}
			schema.Properties[key] = prop
			if !v.Optional {
				schema.Required = append(schema.Required, key)
			}
		}
	}

	return schema
}

// Validate checks input arguments against a JSON Schema. Returns a list of errors.
func Validate(input map[string]any, schema JsonSchema) []string {
	var errors []string

	for _, key := range schema.Required {
		if input[key] == nil {
			errors = append(errors, fmt.Sprintf("Missing required field: %s", key))
		}
	}

	for key, value := range input {
		prop, ok := schema.Properties[key]
		if !ok {
			continue
		}

		actual := jsonType(value)
		if actual != prop.Type {
			errors = append(errors, fmt.Sprintf("Field %q expected %s, got %s", key, prop.Type, actual))
		}
	}

	return errors
}

func jsonType(v any) string {
	switch v.(type) {
	case string:
		return "string"
	case float64, float32, int, int64, int32:
		return "number"
	case bool:
		return "boolean"
	case []any:
		return "array"
	case map[string]any:
		return "object"
	default:
		return fmt.Sprintf("%T", v)
	}
}

File: java/src/main/java/io/antidrift/zeromcp/Schema.java
package io.antidrift.zeromcp;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Converts simplified Input definitions to JSON Schema and validates arguments.
 */
public final class Schema {

    private Schema() {}

    /**
     * Converts a list of Input fields to a JSON Schema object.
     */
    public static JsonObject toJsonSchema(List<Input> inputs) {
        var schema = new JsonObject();
        schema.addProperty("type", "object");

        var properties = new JsonObject();
        var required = new JsonArray();

        for (var input : inputs) {
            var prop = new JsonObject();
            prop.addProperty("type", input.type().jsonType());
            if (input.description() != null) {
                prop.addProperty("description", input.description());
            }
            properties.add(input.name(), prop);
            if (!input.optional()) {
                required.add(input.name());
            }
        }

        schema.add("properties", properties);
        schema.add("required", required);
        return schema;
    }

    /**
     * Validates arguments against a JSON Schema. Returns a list of error messages.
     */
    public static List<String> validate(Map<String, Object> args, JsonObject schema) {
        var errors = new ArrayList<String>();
        var properties = schema.getAsJsonObject("properties");
        var required = schema.getAsJsonArray("required");

        if (required != null) {
            for (var elem : required) {
                var key = elem.getAsString();
                if (!args.containsKey(key) || args.get(key) == null) {
                    errors.add("Missing required field: " + key);
                }
            }
        }

        if (properties != null) {
            for (var entry : args.entrySet()) {
                var prop = properties.getAsJsonObject(entry.getKey());
                if (prop == null) continue;

                var expectedType = prop.get("type").getAsString();
                var actualType = jsonTypeOf(entry.getValue());
                if (!actualType.equals(expectedType)) {
                    errors.add("Field \"%s\" expected %s, got %s"
                        .formatted(entry.getKey(), expectedType, actualType));
                }
            }
        }

        return errors;
    }

    private static String jsonTypeOf(Object value) {
        if (value == null) return "null";
        if (value instanceof String) return "string";
        if (value instanceof Boolean) return "boolean";
        if (value instanceof Number) return "number";
        if (value instanceof List<?>) return "array";
        if (value instanceof Map<?,?>) return "object";
        if (value instanceof JsonElement je) {
            if (je.isJsonPrimitive()) {
                var p = je.getAsJsonPrimitive();
                if (p.isString()) return "string";
                if (p.isBoolean()) return "boolean";
                if (p.isNumber()) return "number";
            }
            if (je.isJsonArray()) return "array";
            if (je.isJsonObject()) return "object";
        }
        return "object";
    }
}
